File: virtual/script/core/Function.go
package core

type (

	// FunctionParam type
	FunctionParam struct {
		name     string
		typeName string
	}

	// FunctionCode type
	FunctionCode struct {
		source      Any
		impl        Native
		next        *Statement
		alternative *Statement
	}

  // Native type
	Native func(reciever Ref, arguments []Ref, ctx ExecutionContext)

	// Function type
	Function struct {
		Object

		name          string
		parameters    []FunctionParam
		boundReciever Ref
		code          *FunctionCode
		// to be parent for a new variable scope in Apply()
		lexicalScope *VariableScope
	}
)

var functionProto *Object = NewObject(Hash{

	Apply: func(reciever Ref, arguments []Ref) (resut Ref, err Error) {

		return ExecuteFunction(reciever, arguments)
	},

	Call: func(reciever Ref, arguments []Ref) (resut Ref, err Error) {

		return ExecuteFunction(reciever, arguments)
	},

	Bind: func(reciever Ref, arguments []Ref) *Function {
		return NewFunction{
			name:          `Bind[%s]`,
			parameters:    reciever.parameters,
			source:        reciever.code.source,
			boundReciever: arguments[0],
		}

	},
})

// NewFunction constructor
func NewFunction(name String, parameters []FunctionParam, reciever Ref, source Any) (fn *Function) {

	fn := &Function{
		typeCode:      TypeCodeFunction,
		next:          functionProto,
		name:          name,
		parameters:    parameters,
		code:          &FunctionCode{source},
		boundReciever: reciever,
		lexicalScope:  CurrentExecutionContext().scope,
	}

	// to be referred as prototype by each object that newly constructed with this function
	prototype := NewObject(&Hash{`constructor`: fn})
	fn.Set(`prototype`, prototype)

	return
}

// Execute method
func (fn *Function) Execute(reciever Ref, arguments []Ref) (result Ref, err Error) {

	// lazy translate source code into binary executable code
	err := ensureCodeIsTranslated(fn)
	if err {
		return
	}

	ctx := CurrentExecutionContext()

	for op := fn.code; op; {
			op, result, err := op.Execute(ctx)
			if err {
				return
			}
	}
	return
}

// ExecuteFunction method
func ExecuteFunction(fn Any, reciever Ref, arguments []Ref) (result Ref, err Error) {

	if fn.boundRecievier {
		recievier = fn.boundRecievier
	}

	// create a new execution context for this invocation
	// and push it into execution stack
	ctx := CreateAndPushExecutionContext(fn, reciever, arguments)
	defer PopExecutionContext()

	result, err := fn.Execute(reciever, arguments)

	return
}

File: virtual/script/native/function.js
import { EXECUTION_STACK, CurrentVariableScope } from './context.js';
import { $NewObject, $Object, SetProperty, PROTO_PROPERTY } from './object.js';
import { $Tuple, $Map, TYPE_CODES, $String, $undefined } from './_core.js';
import { ensureCodeIsTranslated } from '../core/Translator.js';

const { FUNCTION } = TYPE_CODES;

const NAME_ANONYMOUS = ('anonymous');
const KEY_CONSTRUCTOR = $String('Constructor');

export const FUNCTION_PROTOTYPE = $Object({

  Apply: ApplyFunction,

  Call: ($, This, ...Arguments) => ApplyFunction($, This, Arguments),

  Bind: ($, BoundToThis) => $Function({ ...$, BoundToThis })
});

export function $Function({
    Parameters = [],
    Code,
    Name = NAME_ANONYMOUS,
    BoundToThis,
    // to be parent for a new variable scope in Apply()
    LexicalScope = CurrentVariableScope(),
    // to be referred as prototype by each object that newly constructed with this function
    Prototype = $NewObject()
  }) {

  const $descr = $Map({ __Proto__: PROTO_PROPERTY });
  const $data = $Map({
    Name,
    Parameters,
    BoundToThis,
    Code,
    LexicalScope,
    Prototype
  });

  const $ = $Tuple(FUNCTION, $data, $descr, FUNCTION_PROTOTYPE);

  SetProperty(Prototype, KEY_CONSTRUCTOR, $);

  return $;
}

export function ApplyFunction(Context, Fn, This = $undefined, Arguments = []) {

  // lazy translate source code into binary executable code
  ensureCodeIsTranslated(Fn);

  // create a new execution context for this invocation
  // and push it into execution stack
  let context = Context.Push({
    Fn,
    This: Fn.BoundToThis || This,
    Arguments
  });

  // Evaluate binary code
  for (let index = 0, op = Fn.Body[index], len = Fn.Body.length;
        index < len && !context.Exit;
        op = Fn.Body[index++]) {

    op.apply(context);
  }

  Context.Pop();

  // to provide context.Result outside
  return context.Result;
}
